pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -495,71 +566,115 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 		if identityMatches(set, replicas[i]) && storageMatches(set, replicas[i]) && retentionMatch {
 			continue
 		}
-		// Make a deep copy so we don't mutate the shared cache
-		replica := replicas[i].DeepCopy()
-		if err := ssc.podControl.UpdateStatefulPod(ctx, updateSet, replica); err != nil {
-			return &status, err
-		}
+		actions = append(actions, func() (bool, error) {
```

Comment:
Isn't it problematic in parallel mode that here we read replicas[i] and potentially in parallel we create/update that pod?

Or We call UpdateSetatefulPod when the pod is already changed in Delete/CreateStatefulPod.

Also: Does the if in like 566 makes sense if the object can change as part of the action?

Can we maybe instead of adding 10 actions per single pod add one huge action per pod?

Like wrap entire for body in a single action?
High level it should be OK, because we will be doing work for multiple pods in parallel.

It will be also much easier to review.

Reply:
Refactored as suggested. Makes it look like a much smaller change, thanks!